tasks/Logger.cpp

```diff
@@ -372,48 +343,126 @@ void Logger::snapshot()
+        log(Info) << "File " << currentFile << " already exists. Overwriting existing file." << endlog();
         return true;
     }
 
     if (!_overwrite_existing_files.get() && _auto_timestamp_files.get())
     {
         do {
-            log(Warning) << "Timestamped file " << _current_file.get() << " already exists. Retrying in 1 second." << endlog();
+            log(Warning) << "Timestamped file " << currentFile << " already exists. Retrying in 1 second." << endlog();
             usleep(1*1000000);
-            timestampFile();
-        } while(boost::filesystem::exists(_current_file.get()));
-        log(Warning) << "Writing log to " << _current_file.get() << " instead." << endlog();
+            currentFile = timestampFile(file);
+        } while(boost::filesystem::exists(currentFile));
+        log(Warning) << "Writing log to " << currentFile << " instead." << endlog();
         return true;
     }
 }
+
+bool Logger::computeCurrentFile(std::string const& file, std::string &currentFile) const
+{
+    if(file.empty())
+    {
+        log(Error) << "Could not create log file. Task property _file is empty."
+                   << endlog();
+        return false;
+    }
+
+    if (_overwrite_existing_files.get() && _auto_timestamp_files.get())
+    {
+        log(Error) << "The properties overwrite_existing_files and "
+                      "auto_timestamp_files are both set to true, "
+                      "but are mutually exclusive." << endlog();
+        return false;
+    }
+
+    currentFile = file;
+
+    if (_auto_timestamp_files.get()) {
+        currentFile = timestampFile(file);
+    }
+
+    if (boost::filesystem::exists(currentFile))
+    {
+        if (!handleExistingFile(file, currentFile))
+        {
+            return false;
+        }
+    }
+    else if (_auto_timestamp_files.get())
+    {
+        log(Info) << "Successfully timestamped log file. "\
+                     "Writing to: " << currentFile << endlog();
+    }
+    return true;
+}
+
+void Logger::updateLoggers(std::auto_ptr<std::ofstream> &io)
+{
+    if (m_io && m_io->is_open()){
+        m_io->close();
+    }
+    // The registry has been loaded on construction
+    // Now, create the output file
+    auto_ptr<Logfile> file(new Logfile(*io));
+
+    for (Reports::iterator it = root.begin(); it != root.end(); ++it)
+    {
+        delete (it->logger);
+        it->logger = new Logging::StreamLogger(
+                it->name, it->type_name, it->cxx_type_name,*(it->registry), it->metadata, *file);
+    }
+
+    delete m_file;
+    m_file = file.release();
+    delete m_io;
+    m_io   = io.release();
+}
+
+bool Logger::setFile(std::string const &value)
+{
+    std::string currentFile;
+    if (!computeCurrentFile(value, currentFile)){
+        return false;
+    }
+    try {
+        auto_ptr<ofstream> io(new ofstream(currentFile.c_str()));
+        updateLoggers(io);
+        _current_file.set(currentFile);
```

Comment:
Move these two lines out of the catch block. They can't throw ofstream::failure.